[3.x] Support empty string markers for translations

Ovnuniarchos opened the pull request:
This patch is for the 3.x branch.

**TranslationServer now allows for the string "<EMPTY STRING>" (sans qutoes) to be converted to an empty string on tr().**
This allows for texts that need translation to some languages, but not others, using a string that the translated text must be equal to. A translation _containing_ the marker _does not_ get translated to an empty string.

**Use case:** An explanation for a term that is self-evident in a language, but is converted to an abbreviation or an acronym in another. The first language doesn't need an explanation, thus its translated string is rendered empty. The second one, on the other hand, needs an explanation to be present.

I think workarounds like using some kind of space may be too reliant on the font used (especially for zero-width spaces), and make empty tooltips appear when used as a tooltip text.

**Concrete example:**
A label with the text "Key on", when translated to Spanish, results in something like "Activación nota". This text doesn't fit in the UI, so it must be shortened to "Act. n.". As this text is too ambiguous, it needs a tooltip explaining its meaning. In English, the text doesn't need any more explanation, so it doesn't need a tooltip.

**Tooltips don't show if their _translated_ text is an empty string.**
Relating to the previous example, the "Key on" label doesn't need a tooltip, so it gets translated to an empty string. But its property "text" is not an empty string so an empty tooltip appears. This patch fixes that.

If this patch is merged I'll try to adapt it to the translation server in 4.x.

- *Production edit: This addresses https://github.com/godotengine/godot-proposals/issues/9139 for 3.x.*

AThousandShips committed:
[Editor] Add `EditorPlugin::scene_saved` signal (3.x backport)

Matches the `EditorNode` one for parity with the exposed
`resource_saved` signal

3.x backport: Adjusted documentation slightly, backported to 3.x branch

Co-authored-by: t20kdc <asdd2808@gmail.com>

Ovnuniarchos committed:
TranslationServer now allows for the string "<EMPTY STRING>" (unquoted) to be converted to an empty string on tr().

Tooltips don't show if their _translated_ text is an empty string.

mhilbrunner committed:
Fix VS project generation with SCons 4.8.0+

(cherry picked from commit f682406cf26618d926ed33d7fd43e93de0348d85)

AThousandShips commented:
This should normally be implemented in 4.x first, and *then* in 3.x if it's considered appropriate to backport

Calinou reviewed (commented):
I wouldn't make the empty string marker string configurable. By making it configurable, it means the empty string marker is no longer standard, which limits reusability of translation files across projects. It also makes it harder to document the setup process for users new to Godot's localization system.

I'd try to figure out a standard value that is used by other translation systems out there. If there isn't any standard in place for this, I suggest something like `#empty#` that is *very* unlikely to be used in existing translations (to avoid breaking existing projects).

Ovnuniarchos commented:
I haven't found any standard for "empty string". In fact, most of the time the matter is brought up, the most accepted solution seems to be "why would you want to do that?".

I will put a fixed, obvious marker. Something like "<EMPTY-STRING>". The fact that the detection is by full string equality, instead of substring equality, simplifies things.

Regarding the 4.x patch: The translation system in 4.x is _much_ more complex than the 3.x one and I'd prefer to study it well beforehand.

aaronfranke committed:
[3.x] Bind Array and Pool*Array get and set functions

jitspoe committed:
Fix random multithreaded crash that happens when setting the audio stream on a AudioStreamRandomPitch stream.

Capital-EX committed:
Backport 65910 for Godot 3

jitspoe committed:
[3.x] Do not auto add default script and country codes to the locale.

timothyqiu committed:
Cache results for `TranslationServer.compare_locales()`

timothyqiu committed:
Fix button click detection when `Tree` is rotated

timothyqiu committed:
Fix `PopupMenu` margin and separation calculations

timothyqiu committed:
Fix Button not listing `hover_pressed` stylebox

kilian-diener committed:
Fix error when non-ASCII characters in resource pack path

Co-authored-by: Haoyu Qiu <timothyqiu32@gmail.com>

akien-mga committed:
certs: Sync with Mozilla bundle as of Oct 19, 2024

https://github.com/bagder/ca-bundle/commit/4d3fe6683f651d96be1bbef316b201e9b33b274d

Document matching mozilla-release changeset.

(cherry picked from commit eb41ff0feeeee9f72127534e10810c3334724702)

akien-mga committed:
CI: Update Ubuntu runners to 24.04, but keep 22.04 for Linux builds

Pin clang-format to version 16, and black to 24.10.0.
Keep using Ubuntu 22.04 for Linux builds for portability.

akien-mga committed:
CI: Remove now unused sources.list file

AThousandShips committed:
[JavaScript] Don't cache emsdk

Due to how caches are accessed this cache is almost useless, it only
matters if it is from the same branch or a base branch, and is identical
between branches, so caching it just clutters the build cache

AThousandShips committed:
[Buildsystem] Improve cache handling

arkology committed:
[3.x] Document Timer autostart in tool scripts

lawnjelly committed:
Ameliorate performance regression due to directional shadow `fade_start`

Calinou committed:
Hide last DirectionalLight shadow split distance property when using PSSM 3 Splits

The last shadow split property only has an effect when using PSSM 4 Splits.

Chubercik committed:
Update the `Thirdparty` section of `CHANGELOG.md`

arkology committed:
Document radial_center_offset bounds

akien-mga committed:
mbedtls: Update to upstream version 2.28.9

(cherry picked from commit 881645fff9b8bb50a2078f52b4223f5ccc3d1180)

twrightsman committed:
Enable builds with miniupnpc API 18

Backports https://github.com/godotengine/godot/pull/97139 without
bumping the embedded miniupnpc library.

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

lawnjelly committed:
Revert changes to Range::set_value #65101

lawnjelly committed:
Physics Interpolation - Auto-reset on `set_physics_interpolation_mode()`

Fixes historical bug where auto-reset wasn't working correctly.
Also fixes process modes on Cameras when mode is changed.

akien-mga committed:
embree: Fix invalid output operators raising errors with GCC 15

lawnjelly committed:
Physics Interpolation - fix client interpolation pump

Client interpolation pump is moved AFTER the physics tick, after physics objects have been moved.
This is necessary because the `current` transform is also updated during the pump.

axewizardbgg committed:
Docs: BitMap: Added description for opaque_to_polygons method
